Algorithm.CSharp/CoarseSelectionTimeRegressionAlgorithm.cs

```diff
@@ -0,0 +1,124 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Data;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Interfaces;
+
+namespace QuantConnect.Algorithm.CSharp
+{
+    /// <summary>
+    /// Test algorithm that reproduces GH issues 3410 and 3409.
+    /// Coarse universe selection should start from the algorithm start date.
+    /// Data returned by history requests performed from the selection method should be up to date.
+    /// </summary>
+    public class CoarseSelectionTimeRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
+    {
+        private Symbol _spy;
+        private decimal _historyCoarseSpyPrice;
+
+        /// <summary>
+        /// Initialise the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must initialized.
+        /// </summary>
+        public override void Initialize()
+        {
+            SetStartDate(2014, 03, 25);
+            SetEndDate(2014, 04, 01);
+
+            _spy = AddEquity("SPY", Resolution.Daily).Symbol;
+
+            UniverseSettings.Resolution = Resolution.Daily;
+            AddUniverse(CoarseSelectionFunction);
+        }
+
+        public IEnumerable<Symbol> CoarseSelectionFunction(IEnumerable<CoarseFundamental> coarse)
+        {
+            var sortedByDollarVolume = coarse.OrderByDescending(x => x.DollarVolume);
+            var top = sortedByDollarVolume
+                .Where(fundamental => fundamental.Symbol != _spy) // ignore spy
+                .Take(1);
+
+            _historyCoarseSpyPrice = History(_spy, 1).First().Close;
+
+            return top.Select(x => x.Symbol);
+        }
+
+        /// <summary>
+        /// OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
+        /// </summary>
+        /// <param name="data">Slice object keyed by symbol containing the stock data</param>
+        public override void OnData(Slice data)
```

Comment:
Do we want to add a check to confirm the date our universe selection was applied by checking for a non-SPY symbol and comparing w/ an expected date?

Reply:
This is handled by the data/active security count, without this fix it throws on the first call, so I think we are covered 👍